crates/bevy_ecs/src/schedule/set.rs

```diff
@@ -2,57 +2,52 @@ use std::any::TypeId;
 use std::fmt::Debug;
 use std::hash::{Hash, Hasher};
 use std::marker::PhantomData;
-use std::sync::atomic::{AtomicUsize, Ordering};
 
 pub use bevy_ecs_macros::{ScheduleLabel, SystemSet};
-use bevy_utils::define_boxed_label;
-use bevy_utils::label::DynHash;
+use bevy_utils::define_label;
+use bevy_utils::intern::Interned;
+pub use bevy_utils::label::DynEq;
 
 use crate::system::{
     ExclusiveSystemParamFunction, IsExclusiveFunctionSystem, IsFunctionSystem, SystemParamFunction,
 };
 
-define_boxed_label!(ScheduleLabel);
-
-/// A shorthand for `Box<dyn SystemSet>`.
-pub type BoxedSystemSet = Box<dyn SystemSet>;
-/// A shorthand for `Box<dyn ScheduleLabel>`.
-pub type BoxedScheduleLabel = Box<dyn ScheduleLabel>;
-
-/// Types that identify logical groups of systems.
-pub trait SystemSet: DynHash + Debug + Send + Sync + 'static {
-    /// Returns `Some` if this system set is a [`SystemTypeSet`].
-    fn system_type(&self) -> Option<TypeId> {
-        None
-    }
-
-    /// Returns `true` if this system set is an [`AnonymousSet`].
-    fn is_anonymous(&self) -> bool {
-        false
-    }
-    /// Creates a boxed clone of the label corresponding to this system set.
-    fn dyn_clone(&self) -> Box<dyn SystemSet>;
-}
-
-impl PartialEq for dyn SystemSet {
-    fn eq(&self, other: &Self) -> bool {
-        self.dyn_eq(other.as_dyn_eq())
-    }
-}
-
-impl Eq for dyn SystemSet {}
-
-impl Hash for dyn SystemSet {
-    fn hash<H: Hasher>(&self, state: &mut H) {
-        self.dyn_hash(state);
+define_label!(
+    /// A strongly-typed class of labels used to identify an [`Schedule`].
+    ScheduleLabel,
+    SCHEDULE_LABEL_INTERNER
+);
+
+define_label!(
+    /// Types that identify logical groups of systems.
+    SystemSet,
+    SYSTEM_SET_INTERNER,
+    {
+        /// Returns `Some` if this system set is a [`SystemTypeSet`].
+        fn system_type(&self) -> Option<TypeId> {
+            None
+        }
```

Comment:
```suggestion
define_label!(
    /// Types that identify logical groups of systems.
    SystemSet,
    SYSTEM_SET_INTERNER,
    extra_methods: {
        /// Returns `Some` if this system set is a [`SystemTypeSet`].
        fn system_type(&self) -> Option<TypeId> {
            None
        }
```

This suggestion won't compile as-is, but to make it more clear what's happening here, can you add names to these blocks like `extra_methods` and `extra_methods_impl`?

Reply:
Done. Also rebased the PR on main again.